File: KotlinLab/KotlinLab/src/main/kotlin/ch2/section7_nullsafety/Test2.kt
package org.example.ch2.section7_nullsafety

fun main() {
    //일반 객체 테스트...
    //==, === 차이 없다.. 무조건 객체(주소값)비교..
    class User
    val user1: User = User()
    val user2: User? = user1
    println("user1 == user2 : ${user1 == user2}")//user1 == user2 : true
    println("user1 === user2 : ${user1 === user2}")//user1 === user2 : true

    val data1: Int = 1000
    val data2: Int = 1000
    val data3: Int? = 1000
    val data4: Int? = 1000
    //기초타입, non-null, 차이없다.. 모두 값 비교
    println("data1 == data2 : ${data1 == data2}")//true
    println("data1 === data2 : ${data1 === data2}")//true

    //기초타입, nullable, == 값, === 객체비교..
    //자바 변형시.. Integer 객체로 변경되기 때문.. 이를 boxing 이라고 한다..
    println("data3 == data4 : ${data3 == data4}")//true
    println("data3 === data4 : ${data3 === data4}")//false
}

File: flutter_lab2/lib/ch2/section4/test2.dart
int data1 = 10;
int? data2 = 10;

//초기화 시점을 미루는...
late int a1;

String? str = "hello";

main(){
  // data1 = data2;//error..
  data1 = data2 as int;//nullable -> non-null : 명시적 캐스팅..
  data2 = data1;//non-null -> nullable : 암시적 캐스팅...

  // print(a1 + 10);//error... LateInitializationError: Field 'a1' has not been initialized.
  a1 = 10;
  print(a1 + 10);

  // str.isEmpty;//error.. nullable 변수를 null safety operator 를 사용하지 않고 이용..
  bool? result = str?.isEmpty;

  int? data3;
  data3 ??= 10;
  print(data3);//10
  data3 ??= null;//대입되는 값이 null 이면 대입이 안된다..
  print(data3);//10

  String? data4 = "hello";
  String? result2 = data4 ?? "world";
  print(result2);//hello
  data4 = null;
  result2 = data4 ?? "world";
  print(result2);//world
}